libs/db/boltdb.go

```diff
@@ -0,0 +1,253 @@
+package db
+
+import (
+	"bytes"
+	"path/filepath"
+	"sync"
+
+	"github.com/etcd-io/bbolt"
+)
+
+var bucket = []byte("boltdb_bucket")
+
+func init() {
+	registerDBCreator(BoltDBBackend, func(name string, dir string) (DB, error) {
+		return NewBoltdb(name, dir)
+	}, false)
+}
+
+type Boltdb struct {
+	db *bbolt.DB
+}
+
+func NewBoltdb(name, dir string) (DB, error) {
+	return NewBoltdbWithOpts(name, dir, bbolt.DefaultOptions)
+}
+
+func NewBoltdbWithOpts(name, dir string, opts *bbolt.Options) (DB, error) {
+	dbPath := filepath.Join(dir, name+".db")
+	db, err := bbolt.Open(dbPath, 0600, opts)
+	if err != nil {
+		return nil, err
+	}
+	err = db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(bucket)
+		return err
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &Boltdb{db: db}, nil
+}
+
+func (bdb *Boltdb) Get(key []byte) (value []byte) {
+	key = nonNilBytes(key)
+	err := bdb.db.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket(bucket)
+		value = b.Get(key)
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	return
+}
+
+func (bdb *Boltdb) Has(key []byte) bool {
+	return bdb.Get(key) != nil
+}
+
+func (bdb *Boltdb) Set(key, value []byte) {
+	key = nonNilBytes(key)
+	value = nonNilBytes(value)
+	err := bdb.db.Update(func(tx *bbolt.Tx) error {
+		b := tx.Bucket(bucket)
+		return b.Put(key, value)
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (bdb *Boltdb) SetSync(key, value []byte) {
+	bdb.Set(key, value)
+}
+
+func (bdb *Boltdb) Delete(key []byte) {
+	key = nonNilBytes(key)
+	err := bdb.db.Update(func(tx *bbolt.Tx) error {
+		return tx.Bucket(bucket).Delete(key)
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (bdb *Boltdb) DeleteSync(key []byte) {
+	bdb.Delete(key)
+}
+
+func (bdb *Boltdb) Close() {
+	bdb.db.Close()
+}
+
+func (bdb *Boltdb) Print() {
+	panic("boltdb.print not yet implemented")
```

Comment:
If we merge this, we should open follow-up issues for these.

Reply:
I meant that if these were not implemented here, we should track these in an issue and do implement these functions in a follow-up PR.